### PR TITLE
Add support for Spectrum Applications

### LINKED ISSUE #207: Add support for Spectrum API
Hi,

could you add support to configure spectrum?

https://developers.cloudflare.com/spectrum/api-reference/

Thanks!

### PR DESCRIPTION
**Pull request by jacobbednarz**
Implements support for managing Spectrum Applications via the API.

API reference: https://developers.cloudflare.com/spectrum/api-reference/

Fixes #207

### CODE REVIEW

**Comment by jacobbednarz** [comment]
Pending getting full access to Spectrum features to test the request/responses.

**Comment by patryk** [comment]
Let me know if you need help getting features enabled for testing.

**Comment by jacobbednarz** [comment]
Appreciate the offer @patryk! Our account manager has come through for us and managed to get it all hooked up.

This should be ready for some 👀 to review it now.

**Comment by jacobbednarz** [comment]
@T4cC0re seeing how you opened the original issue, are you able to please check this out and ensure it covers all your needs?

**Comment by T4cC0re** [comment]
Sure thing! Will do tomorrow. Thanks so far!

**Comment by T4cC0re** [comment]
Hi, just wanted to let you know, that I did not have time to check it, yet. Will do tomorrow (for real this time).

**Comment by T4cC0re** [comment]
Hi, so for me it does not quite work :/

```
log.Infoln(app.OriginDirect)
direct := fmt.Sprintf("tcp://%s:1337", (*data).ExternalIP)
AppId := app.ID
app.OriginDirect = []string{direct}
app.ID = ""
log.Infoln(app.OriginDirect)
_, err :=api.UpdateSpectrumApplication(zoneID, AppId, app)
if err != nil {
	log.Errorln(err)
}
```

The content of `app.OriginDirect` is:
```
time="2018-08-02T20:37:42Z" level=info msg="[tcp://1.2.3.4:1337]" source="hook.go:122"
time="2018-08-02T20:37:42Z" level=info msg="[tcp://5.6.7.8:1337]" source="hook.go:127"
```
The error is:
```
time="2018-08-02T20:37:42Z" level=error msg="error from makeRequest: HTTP status 400: content \"{\\n  \\\"result\\\": null,\\n  \\\"success\\\": false,\\n  \\\"errors\\\": [\\n    {\\n      \\\"code\\\": 11000,\\n      \\\"message\\\": \\\"proxyanything.api.error.invalid_origin_configuration\\\"\\n    }\\n  ],\\n  \\\"messages\\\": []\\n}\\n\"" source="hook.go:130"
```
I set the `app.ID` to an empty value, because the doc says so.

Let me check really quick if the error disappears If I add `omitempty` to the json field exports. Because the doc says that an update will fail if multiple origin configs are set.

**Comment by T4cC0re** [comment]
I did the following:
```
sed -i 's/json:\(.*\)"`$/json:\1,omitempty"`/gmi'  spectrum.go
```
Which added `omitempty` to each json field export, but that did not help. I get the same error message.

**Comment by T4cC0re** [comment]
@jacobbednarz can you look into it?

**Comment by jacobbednarz** [comment]
I'm on vacation for another week or so but can take a look once I'm back.

________________________________

**Comment by jacobbednarz** [comment]
I've had a quick look at this today and the issue is that the API is getting both `origin_direct` and `origin_dns` which causes a validation error [per the documentation](https://developers.cloudflare.com/spectrum/api-reference/#update-a-spectrum-application) due to not using a pointer for the `SpectrumApplicationOriginDNS` struct.

I've pushed up 4980932 which should address that if you'd like to give it another whirl.

**Comment by T4cC0re** [comment]
Yep, that specific case now works!
Will test more soon-ish (Create and different updates)

**Comment by jacobbednarz** [comment]
@T4cC0re unless you have any further issues, I think this is ready for merging @patryk.

**Comment by T4cC0re** [comment]
Hi @jacobbednarz 
I get an error during creation of a Spectrum App. Will see what the exact problem is, might be my code.

```
                        direct := fmt.Sprintf("tcp://%s:%d", hookData.ExternalIP, port)
			proto := fmt.Sprintf("tcp/%d", port)

			app := cloudflare.SpectrumApplication{
				DNS: cloudflare.SpectrumApplicationDNS{
					Name: dns,
					Type: "A",
				},
				OriginDirect:  []string{direct},
				IPv4:          true,
				ProxyProtocol: false,
				Protocol:      proto,
			}
			_, err := api.CreateSpectrumApplication(zoneID, app)
			if err != nil {
				log.Error(err)
			}
```

Error is:

```
"code": 11004,
"message": "The DNS configuration is invalid"
```

**Comment by T4cC0re** [comment]
Nevermind. DNS.Type has to be "CNAME"

Works from my PoV

**Comment by jacobbednarz** [comment]
@patryk we're 👌 to :ship: here

File: spectrum.go [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+type SpectrumApplicationOriginDNS struct {
+	Type  string   `json:"type"`
+	Names []string `json:"names"`
+	TTL   int      `json:"ttl"`
+}
+
+// SpectrumApplicationDetailResponse is the structure of the detailed response
+// from the API.
+type SpectrumApplicationDetailResponse struct {
+	Success  bool                `json:"success"`
```
Comment (patryk) [inline spectrum.go]: Could we reuse `Response` struct here (and other structs we add in this PR)?

File: spectrum.go [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+type SpectrumApplicationsDetailResponse struct {
+	Success  bool                  `json:"success"`
+	Errors   []string              `json:"errors"`
+	Messages []string              `json:"messages"`
+	Result   []SpectrumApplication `json:"result"`
+}
+
+// DeletedSpectrumApplicationResponse defines the API response when deleting a
+// Spectrum application.
+type DeletedSpectrumApplicationResponse struct {
```
Comment (patryk) [inline spectrum.go]: Not sure we actually need a specific struct type here, just to catch the ID we know already. Other delete methods usually don't return anything.

What's your opinion?

File: spectrum.go [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+		return DeletedSpectrumApplicationResult{}, errors.Wrap(err, errMakeRequestError)
+	}
+
+	var deletedSpectrumApplication DeletedSpectrumApplicationResponse
+	err = json.Unmarshal(res, &deletedSpectrumApplication)
+	if err != nil {
+		return DeletedSpectrumApplicationResult{}, errors.Wrap(err, errUnmarshalError)
+	}
+
+	return deletedSpectrumApplication.Result, nil
```
Comment (patryk) [inline spectrum.go]: We don't need to return anything here (besides errors). No useful details in API result.

File: spectrum.go (hunk 2 above)
Comment (jacobbednarz) [reply spectrum.go]: This was only to support the `Delete` method which is needed. I'll get rid of this.

File: spectrum.go (hunk 3 above)
Comment (jacobbednarz) [reply spectrum.go]: You're spot on. This was left over from testing another part of the SDK using spectrum but it can go like the others.

**Comment by jacobbednarz** [comment]
I'm not sure what's caused it but I can't push to my fork and have it reflected in this PR. Looks like a GitHub issue 🤔 



...so here's the patch instead. Do you mind applying this directly onto this branch for me?

```patch
From da5280ec1a5e2123c7ea2475214d16dbe88d56bc Mon Sep 17 00:00:00 2001
From: Jacob Bednarz <jacob.bednarz@gmail.com>
Date: Wed, 14 Nov 2018 12:06:20 +1100
Subject: [PATCH 1/2] Make `Delete` consistent with other resources

The delete functions don't need to return the resource, just the error
if it's not `nil`.
---
 spectrum.go      | 14 ++++----------
 spectrum_test.go | 10 ++--------
 2 files changed, 6 insertions(+), 18 deletions(-)

diff --git a/spectrum.go b/spectrum.go
index 5887530..afbea90 100644
--- a/spectrum.go
+++ b/spectrum.go
@@ -163,23 +163,17 @@ func (api *API) UpdateSpectrumApplication(zoneID, appID string, appDetails Spect
 // DeleteSpectrumApplication removes a Spectrum application based on the ID.
 //
 // API reference: https://developers.cloudflare.com/spectrum/api-reference/#delete-a-spectrum-application
-func (api *API) DeleteSpectrumApplication(zoneID string, applicationID string) (DeletedSpectrumApplicationResult, error) {
+func (api *API) DeleteSpectrumApplication(zoneID string, applicationID string) error {
 	uri := fmt.Sprintf(
 		"/zones/%s/spectrum/apps/%s",
 		zoneID,
 		applicationID,
 	)
 
-	res, err := api.makeRequest("DELETE", uri, nil)
+	_, err := api.makeRequest("DELETE", uri, nil)
 	if err != nil {
-		return DeletedSpectrumApplicationResult{}, errors.Wrap(err, errMakeRequestError)
+		return errors.Wrap(err, errMakeRequestError)
 	}
 
-	var deletedSpectrumApplication DeletedSpectrumApplicationResponse
-	err = json.Unmarshal(res, &deletedSpectrumApplication)
-	if err != nil {
-		return DeletedSpectrumApplicationResult{}, errors.Wrap(err, errUnmarshalError)
-	}
-
-	return deletedSpectrumApplication.Result, nil
+	return nil
 }
diff --git a/spectrum_test.go b/spectrum_test.go
index 19ca5e5..62ff1c4 100644
--- a/spectrum_test.go
+++ b/spectrum_test.go
@@ -257,12 +257,6 @@ func TestDeleteSpectrumApplication(t *testing.T) {
 
 	mux.HandleFunc("/zones/01a7362d577a6c3019a474fd6f485823/spectrum/apps/f68579455bd947efb65ffa1bcf33b52c", handler)
 
-	want := DeletedSpectrumApplicationResult{
-		ID: "40d67c87c6cd4b889a4fd57805225e85",
-	}
-
-	actual, err := client.DeleteSpectrumApplication("01a7362d577a6c3019a474fd6f485823", "f68579455bd947efb65ffa1bcf33b52c")
-	if assert.NoError(t, err) {
-		assert.Equal(t, want, actual)
-	}
+	err := client.DeleteSpectrumApplication("01a7362d577a6c3019a474fd6f485823", "f68579455bd947efb65ffa1bcf33b52c")
+	assert.NoError(t, err)
 }
-- 
2.19.1


From 39096834769d5da13b49c6e7ea59d5f2b39f059c Mon Sep 17 00:00:00 2001
From: Jacob Bednarz <jacob.bednarz@gmail.com>
Date: Wed, 14 Nov 2018 12:06:38 +1100
Subject: [PATCH 2/2] Remove duplicate fields in favour of `Response`

---
 spectrum.go | 27 ++++-----------------------
 1 file changed, 4 insertions(+), 23 deletions(-)

diff --git a/spectrum.go b/spectrum.go
index afbea90..08d0825 100644
--- a/spectrum.go
+++ b/spectrum.go
@@ -42,34 +42,15 @@ type SpectrumApplicationOriginDNS struct {
 // SpectrumApplicationDetailResponse is the structure of the detailed response
 // from the API.
 type SpectrumApplicationDetailResponse struct {
-	Success  bool                `json:"success"`
-	Errors   []string            `json:"errors"`
-	Messages []string            `json:"messages"`
-	Result   SpectrumApplication `json:"result"`
+	Response
+	Result SpectrumApplication `json:"result"`
 }
 
 // SpectrumApplicationsDetailResponse is the structure of the detailed response
 // from the API.
 type SpectrumApplicationsDetailResponse struct {
-	Success  bool                  `json:"success"`
-	Errors   []string              `json:"errors"`
-	Messages []string              `json:"messages"`
-	Result   []SpectrumApplication `json:"result"`
-}
-
-// DeletedSpectrumApplicationResponse defines the API response when deleting a
-// Spectrum application.
-type DeletedSpectrumApplicationResponse struct {
-	Success  bool                             `json:"success"`
-	Errors   []string                         `json:"errors"`
-	Messages []string                         `json:"messages"`
-	Result   DeletedSpectrumApplicationResult `json:"result"`
-}
-
-// DeletedSpectrumApplicationResult defines the struct for a deleted application
-// result.
-type DeletedSpectrumApplicationResult struct {
-	ID string `json:"id"`
+	Response
+	Result []SpectrumApplication `json:"result"`
 }
 
 // SpectrumApplications fetches all of the Spectrum applications for a zone.
-- 
2.19.1
```

File: spectrum.go (hunk 1 above)
Comment (jacobbednarz) [reply spectrum.go]: Yep. Have updated in the patch. Thanks for picking it up!

**Comment by patryk** [comment]
That's interesting. I cannot do it as well. Heh. I'll merge without change and push commit manually as git email patch.

**Comment by jacobbednarz** [comment]
Nice one! 🌟 Thanks!

**Pull request merged**